usr/src/uts/common/fs/zfs/zap.c

```diff
@@ -575,7 +575,13 @@ zap_deref_leaf(zap_t *zap, uint64_t h, dmu_tx_t *tx, krw_t lt, zap_leaf_t **lp)
 
 	ASSERT(zap->zap_dbuf == NULL ||
 	    zap_f_phys(zap) == zap->zap_dbuf->db_data);
-	ASSERT3U(zap_f_phys(zap)->zap_magic, ==, ZAP_MAGIC);
+
+	/* Reality check for corrupt zap leaves. */
+	if (zap_f_phys(zap)->zap_block_type != ZBT_LEAF ||
+	    zap_f_phys(zap)->zap_magic != ZAP_MAGIC) {
```

Comment:
This also needs to check for block_type == ZBT_HEADER.  Either that, or my rpool is VERY screwed up.  :)